// @ai-hero/sandcastle/dist/AgentProvider.js
const shellEscape = (s) => "'" + s.replace(/'/g, "'\\''") + "'";
/** Maps allowlisted tool names to the input field containing the display arg */
const TOOL_ARG_FIELDS = {
    Bash: "command",
    WebSearch: "query",
    WebFetch: "url",
    Agent: "description",
};
/**
 * Extract an error message from a parsed JSON error event.
 * Handles { error: "string" }, { error: { message: "string" } }, and { message: "string" }.
 */
const extractErrorMessage = (obj) => {
    const err = obj.error;
    if (typeof err === "string")
        return err;
    if (typeof err === "object" && err !== null && typeof err.message === "string") {
        return err.message;
    }
    if (typeof obj.message === "string")
        return obj.message;
    return undefined;
};
const parseStreamJsonLine = (line) => {
    if (!line.startsWith("{"))
        return [];
    try {
        const obj = JSON.parse(line);
        if (obj.type === "assistant" && Array.isArray(obj.message?.content)) {
            const events = [];
            const texts = [];
            for (const block of obj.message.content) {
                if (block.type === "text" && typeof block.text === "string") {
                    texts.push(block.text);
                }
                else if (block.type === "tool_use" &&
                    typeof block.name === "string" &&
                    block.input !== undefined) {
                    const argField = TOOL_ARG_FIELDS[block.name];
                    if (argField === undefined)
                        continue; // not allowlisted
                    const argValue = block.input[argField];
                    if (typeof argValue !== "string")
                        continue; // missing/wrong arg field
                    if (texts.length > 0) {
                        events.push({ type: "text", text: texts.join("") });
                        texts.length = 0;
                    }
                    events.push({
                        type: "tool_call",
                        name: block.name,
                        args: argValue,
                    });
                }
            }
            if (texts.length > 0) {
                events.push({ type: "text", text: texts.join("") });
            }
            return events;
        }
        if (obj.type === "result" && typeof obj.result === "string") {
            return [{ type: "result", result: obj.result }];
        }
        if (obj.type === "system" &&
            obj.subtype === "init" &&
            typeof obj.session_id === "string") {
            return [{ type: "session_id", sessionId: obj.session_id }];
        }
    }
    catch {
        // Not valid JSON — skip
    }
    return [];
};
export const DEFAULT_MODEL = "claude-opus-4-6";
// ---------------------------------------------------------------------------
// Pi agent provider
// ---------------------------------------------------------------------------
const parsePiStreamLine = (line) => {
    if (!line.startsWith("{"))
        return [];
    try {
        const obj = JSON.parse(line);
        if (obj.type === "message_update" && obj.assistantMessageEvent) {
            const evt = obj.assistantMessageEvent;
            if (evt.type === "text_delta" && typeof evt.delta === "string") {
                return [{ type: "text", text: evt.delta }];
            }
            return [];
        }
        if (obj.type === "tool_execution_start") {
            const toolName = obj.toolName;
            if (typeof toolName !== "string")
                return [];
            const argField = TOOL_ARG_FIELDS[toolName];
            if (argField === undefined)
                return [];
            const args = obj.args;
            if (!args)
                return [];
            const argValue = args[argField];
            if (typeof argValue !== "string")
                return [];
            return [{ type: "tool_call", name: toolName, args: argValue }];
        }
        // Pi emits agent_error / error events on stdout (not stderr) for auth
        // failures, rate limits, and API errors. Capture them as result events so
        // the Orchestrator's stderr-empty fallback can surface them to the user.
        if (obj.type === "agent_error" || obj.type === "error") {
            const msg = extractErrorMessage(obj);
            return msg ? [{ type: "result", result: msg }] : [];
        }
        if (obj.type === "agent_end" && Array.isArray(obj.messages)) {
            const messages = obj.messages;
            for (let i = messages.length - 1; i >= 0; i--) {
                const msg = messages[i];
                if (msg?.role === "assistant") {
                    const texts = [];
                    for (const block of msg.content) {
                        if (block.type === "text" && typeof block.text === "string") {
                            texts.push(block.text);
                        }
                    }
                    if (texts.length > 0) {
                        return [{ type: "result", result: texts.join("") }];
                    }
                    break;
                }
            }
            return [];
        }
    }
    catch {
        // Not valid JSON — skip
    }
    return [];
};
export const pi = (model, options) => ({
    name: "pi",
    env: options?.env ?? {},
    captureSessions: false,
    buildPrintCommand({ prompt }) {
        return {
            command: `pi -p --mode json --no-session --model ${shellEscape(model)}`,
            stdin: prompt,
        };
    },
    buildInteractiveArgs({ prompt }) {
        const args = ["pi", "--model", model];
        if (prompt)
            args.push(prompt);
        return args;
    },
    parseStreamLine(line) {
        return parsePiStreamLine(line);
    },
});
// ---------------------------------------------------------------------------
// Codex agent provider
// ---------------------------------------------------------------------------
const parseCodexStreamLine = (line) => {
    if (!line.startsWith("{"))
        return [];
    try {
        const obj = JSON.parse(line);
        // item.completed with agent_message → text + result
        if (obj.type === "item.completed" &&
            obj.item?.type === "agent_message" &&
            typeof obj.item.text === "string") {
            const text = obj.item.text;
            return [
                { type: "text", text },
                { type: "result", result: text },
            ];
        }
        // item.started with command_execution → tool call
        if (obj.type === "item.started" &&
            obj.item?.type === "command_execution" &&
            typeof obj.item.command === "string") {
            return [{ type: "tool_call", name: "Bash", args: obj.item.command }];
        }
        // Codex emits error events on stdout (not stderr) for auth failures,
        // rate limits, and API errors. Capture them as result events so the
        // Orchestrator's stderr-empty fallback can surface them to the user.
        if (obj.type === "error") {
            const msg = extractErrorMessage(obj);
            return msg ? [{ type: "result", result: msg }] : [];
        }
        // turn.completed → skip
    }
    catch {
        // Not valid JSON — skip
    }
    return [];
};
export const codex = (model, options) => ({
    name: "codex",
    env: options?.env ?? {},
    captureSessions: false,
    buildPrintCommand({ prompt }) {
        const effortFlag = options?.effort
            ? ` -c ${shellEscape(`model_reasoning_effort="${options.effort}"`)}`
            : "";
        return {
            command: `codex exec --json --dangerously-bypass-approvals-and-sandbox -m ${shellEscape(model)}${effortFlag}`,
            stdin: prompt,
        };
    },
    buildInteractiveArgs({ prompt }) {
        const args = ["codex", "--model", model];
        if (prompt)
            args.push(prompt);
        return args;
    },
    parseStreamLine(line) {
        return parseCodexStreamLine(line);
    },
});
export const opencode = (model, options) => ({
    name: "opencode",
    env: options?.env ?? {},
    captureSessions: false,
    buildPrintCommand({ prompt }) {
        const variantFlag = options?.variant
            ? ` --variant ${shellEscape(options.variant)}`
            : "";
        return {
            command: `opencode run --model ${shellEscape(model)}${variantFlag} ${shellEscape(prompt)}`,
        };
    },
    buildInteractiveArgs({ prompt }) {
        const args = ["opencode", "--model", model];
        if (prompt)
            args.push("-p", prompt);
        return args;
    },
    parseStreamLine(_line) {
        return [];
    },
});
export const claudeCode = (model, options) => ({
    name: "claude-code",
    env: options?.env ?? {},
    captureSessions: options?.captureSessions ?? true,
    buildPrintCommand({ prompt, dangerouslySkipPermissions, resumeSession, }) {
        const skipPerms = dangerouslySkipPermissions
            ? " --dangerously-skip-permissions"
            : "";
        const effortFlag = options?.effort ? ` --effort ${options.effort}` : "";
        const resumeFlag = resumeSession
            ? ` --resume ${shellEscape(resumeSession)}`
            : "";
        return {
            command: `claude --print --verbose${skipPerms} --output-format stream-json --model ${shellEscape(model)}${effortFlag}${resumeFlag} -p -`,
            stdin: prompt,
        };
    },
    buildInteractiveArgs({ prompt, dangerouslySkipPermissions, }) {
        const args = ["claude"];
        if (dangerouslySkipPermissions)
            args.push("--dangerously-skip-permissions");
        args.push("--model", model);
        if (options?.effort)
            args.push("--effort", options.effort);
        if (prompt)
            args.push(prompt);
        return args;
    },
    parseStreamLine(line) {
        return parseStreamJsonLine(line);
    },
    parseSessionUsage(content) {
        const lines = content.split("\n");
        for (let i = lines.length - 1; i >= 0; i--) {
            const line = lines[i];
            if (!line.startsWith("{"))
                continue;
            try {
                const obj = JSON.parse(line);
                if (obj.type === "assistant" && obj.message?.usage) {
                    const u = obj.message.usage;
                    if (typeof u.input_tokens === "number" &&
                        typeof u.cache_creation_input_tokens === "number" &&
                        typeof u.cache_read_input_tokens === "number" &&
                        typeof u.output_tokens === "number") {
                        return {
                            inputTokens: u.input_tokens,
                            cacheCreationInputTokens: u.cache_creation_input_tokens,
                            cacheReadInputTokens: u.cache_read_input_tokens,
                            outputTokens: u.output_tokens,
                        };
                    }
                }
            }
            catch {
                // Not valid JSON — skip
            }
        }
        return undefined;
    },
});
