// @ai-hero/sandcastle/dist/createWorktree.js
import { NodeContext, NodeFileSystem } from "@effect/platform-node";
import { existsSync } from "node:fs";
import { join } from "node:path";
import { Effect, Layer } from "effect";
import { hostSessionStore } from "./SessionStore.js";
import { ClackDisplay, Display, FileDisplay } from "./Display.js";
import { preprocessPrompt } from "./PromptPreprocessor.js";
import { resolvePrompt } from "./PromptResolver.js";
import { SandboxFactory, makeSandboxLayerFromHandle, resolveGitMounts, SANDBOX_REPO_DIR, } from "./SandboxFactory.js";
import { withSandboxLifecycle, runHostHooks, } from "./SandboxLifecycle.js";
import { createSandboxFromWorktree } from "./createSandbox.js";
import { buildLogFilename, printFileDisplayStartup } from "./run.js";
import { orchestrate } from "./Orchestrator.js";
import { defaultSessionPathsLayer } from "./SessionPaths.js";
import { callbackAgentStreamEmitterLayer, noopAgentStreamEmitterLayer, } from "./AgentStreamEmitter.js";
import { resolveEnv } from "./EnvResolver.js";
import { mergeProviderEnv } from "./mergeProviderEnv.js";
import { startSandbox } from "./startSandbox.js";
import { syncOut } from "./syncOut.js";
import * as WorktreeManager from "./WorktreeManager.js";
import { copyToWorktree } from "./CopyToWorktree.js";
import { resolveCwd } from "./resolveCwd.js";
import { substitutePromptArgs, validateNoArgsWithInlinePrompt, validateNoBuiltInArgOverride, BUILT_IN_PROMPT_ARG_KEYS, } from "./PromptArgumentSubstitution.js";
import { noSandbox } from "./sandboxes/no-sandbox.js";
import { raceAbortSignal } from "./raceAbortSignal.js";
/**
 * Creates a git worktree as an independent, first-class worktree.
 * Returns a Worktree handle with close() and [Symbol.asyncDispose]().
 *
 * Only accepts 'branch' and 'merge-to-head' strategies — 'head' is a
 * compile-time type error since head means no worktree.
 */
export const createWorktree = async (options) => {
    const branch = options.branchStrategy.type === "branch"
        ? options.branchStrategy.branch
        : undefined;
    const baseBranch = options.branchStrategy.type === "branch"
        ? options.branchStrategy.baseBranch
        : undefined;
    const { hostRepoDir, worktreeInfo } = await Effect.gen(function* () {
        const hostRepoDir = yield* resolveCwd(options.cwd);
        yield* WorktreeManager.pruneStale(hostRepoDir).pipe(Effect.catchAll(() => Effect.void));
        const info = yield* WorktreeManager.create(hostRepoDir, {
            branch,
            baseBranch,
        });
        if (options.copyToWorktree && options.copyToWorktree.length > 0) {
            yield* copyToWorktree(options.copyToWorktree, hostRepoDir, info.path, options.timeouts?.copyToWorktreeMs);
        }
        // Run host.onWorktreeReady hooks after copyToWorktree, before sandbox creation
        if (options.hooks?.host?.onWorktreeReady?.length) {
            yield* runHostHooks(options.hooks.host.onWorktreeReady, info.path);
        }
        return { hostRepoDir, worktreeInfo: info };
    }).pipe(Effect.provide(NodeContext.layer), Effect.runPromise);
    let closed = false;
    const close = async () => {
        if (closed)
            return { preservedWorktreePath: undefined };
        closed = true;
        return Effect.gen(function* () {
            const isDirty = yield* WorktreeManager.hasUncommittedChanges(worktreeInfo.path).pipe(Effect.catchAll(() => Effect.succeed(false)));
            if (isDirty) {
                return { preservedWorktreePath: worktreeInfo.path };
            }
            yield* WorktreeManager.remove(worktreeInfo.path).pipe(Effect.catchAll(() => Effect.void));
            return { preservedWorktreePath: undefined };
        }).pipe(Effect.runPromise);
    };
    const worktreeInteractive = async (opts) => {
        // If signal is already aborted, reject immediately without any setup
        opts.signal?.throwIfAborted();
        const { prompt, promptFile, hooks, agent: provider } = opts;
        const resolvedSandbox = opts.sandbox ?? noSandbox();
        // Validate buildInteractiveArgs is available
        if (!provider.buildInteractiveArgs) {
            throw new Error(`Agent provider "${provider.name}" does not support buildInteractiveArgs, required for interactive sessions.`);
        }
        const inner = Effect.gen(function* () {
            const d = yield* Display;
            // 1. Resolve prompt (from string or file), or skip if neither provided
            const hasPromptSource = prompt !== undefined || promptFile !== undefined;
            const resolved = hasPromptSource
                ? yield* resolvePrompt({ prompt, promptFile })
                : undefined;
            const rawPrompt = resolved?.text ?? "";
            const isInlinePrompt = resolved?.source === "inline";
            // 2. Resolve env vars
            const resolvedEnv = yield* resolveEnv(hostRepoDir);
            const env = mergeProviderEnv({
                resolvedEnv,
                agentProviderEnv: provider.env,
                sandboxProviderEnv: resolvedSandbox.env,
            });
            const effectiveEnv = { ...env, ...(opts.env ?? {}) };
            // 3. Prompt args substitution (skip when no prompt, or when inline passthrough)
            let substitutedPrompt = rawPrompt;
            if (hasPromptSource && !isInlinePrompt) {
                const userArgs = opts.promptArgs ?? {};
                yield* validateNoBuiltInArgOverride(userArgs);
                const effectiveArgs = {
                    SOURCE_BRANCH: worktreeInfo.branch,
                    TARGET_BRANCH: worktreeInfo.branch,
                    ...userArgs,
                };
                const builtInArgKeysSet = new Set(BUILT_IN_PROMPT_ARG_KEYS);
                substitutedPrompt = yield* substitutePromptArgs(rawPrompt, effectiveArgs, builtInArgKeysSet);
            }
            else if (isInlinePrompt) {
                yield* validateNoArgsWithInlinePrompt(opts.promptArgs ?? {});
            }
            // Display intro
            yield* d.intro(opts.name ?? "sandcastle interactive");
            yield* d.summary("Interactive Session", {
                Agent: opts.name ?? provider.name,
                Sandbox: resolvedSandbox.name,
                Branch: worktreeInfo.branch,
            });
            // 4. Start sandbox
            let handle;
            if (resolvedSandbox.tag === "none") {
                handle = yield* Effect.promise(() => resolvedSandbox.create({
                    worktreePath: worktreeInfo.path,
                    env: effectiveEnv,
                }));
            }
            else if (resolvedSandbox.tag === "isolated") {
                const startResult = yield* d.taskLog("Starting sandbox", () => startSandbox({
                    provider: resolvedSandbox,
                    hostRepoDir: worktreeInfo.path,
                    env: effectiveEnv,
                }));
                handle = startResult.handle;
            }
            else {
                const gitPath = join(hostRepoDir, ".git");
                const gitMounts = yield* resolveGitMounts(gitPath);
                const startResult = yield* d.taskLog("Starting sandbox", () => startSandbox({
                    provider: resolvedSandbox,
                    hostRepoDir,
                    env: effectiveEnv,
                    worktreeOrRepoPath: worktreeInfo.path,
                    gitMounts,
                    repoDir: SANDBOX_REPO_DIR,
                }));
                handle = startResult.handle;
            }
            // Run lifecycle — worktree owns worktree, so no worktree cleanup here
            return yield* Effect.gen(function* () {
                if (!handle.interactiveExec) {
                    throw new Error(`Sandbox provider does not support interactiveExec. ` +
                        `The provider must implement the optional interactiveExec method to use interactive().`);
                }
                const interactiveExecFn = handle.interactiveExec.bind(handle);
                const sandboxLayer = makeSandboxLayerFromHandle(handle);
                const worktreePath = handle.worktreePath;
                const applyToHost = resolvedSandbox.tag === "isolated"
                    ? () => syncOut(worktreeInfo.path, handle)
                    : () => Effect.void;
                const lifecycleEffect = withSandboxLifecycle({
                    hostRepoDir,
                    sandboxRepoDir: worktreePath,
                    hooks,
                    branch: worktreeInfo.branch,
                    hostWorktreePath: worktreeInfo.path,
                    applyToHost,
                }, (ctx) => Effect.gen(function* () {
                    const fullPrompt = !hasPromptSource || isInlinePrompt
                        ? substitutedPrompt
                        : yield* preprocessPrompt(substitutedPrompt, ctx.sandbox, ctx.sandboxRepoDir);
                    const interactiveArgs = provider.buildInteractiveArgs({
                        prompt: fullPrompt,
                        dangerouslySkipPermissions: resolvedSandbox.tag !== "none",
                    });
                    const result = yield* raceAbortSignal(Effect.promise(() => interactiveExecFn(interactiveArgs, {
                        stdin: process.stdin,
                        stdout: process.stdout,
                        stderr: process.stderr,
                        cwd: worktreePath,
                    })), opts.signal);
                    return result.exitCode;
                }));
                const lifecycleResult = yield* lifecycleEffect.pipe(Effect.provide(sandboxLayer));
                const exitCode = lifecycleResult.result;
                // Summary
                yield* d.summary("Session Complete", {
                    Commits: String(lifecycleResult.commits.length),
                    Branch: lifecycleResult.branch,
                    "Exit code": String(exitCode),
                });
                return {
                    commits: lifecycleResult.commits,
                    branch: lifecycleResult.branch,
                    preservedWorktreePath: undefined,
                    exitCode,
                };
            }).pipe(
            // Always close sandbox handle
            Effect.ensuring(Effect.promise(() => handle.close().catch(() => { }))));
        });
        try {
            return await Effect.runPromise(inner.pipe(Effect.provide(ClackDisplay.layer), Effect.provide(NodeContext.layer), Effect.provide(NodeFileSystem.layer)));
        }
        catch (error) {
            // If the signal was aborted, surface its reason verbatim (no wrapping)
            opts.signal?.throwIfAborted();
            throw error;
        }
    };
    const worktreeRun = async (opts) => {
        // If signal is already aborted, reject immediately without any setup
        opts.signal?.throwIfAborted();
        const { prompt, promptFile, hooks, agent: provider } = opts;
        const sandboxProvider = opts.sandbox;
        const maxIterations = opts.maxIterations ?? 1;
        if (opts.resumeSession && maxIterations > 1) {
            throw new Error("resumeSession cannot be combined with maxIterations > 1. " +
                "Resume applies to iteration 1 only; multi-iteration resume semantics are not supported.");
        }
        if (opts.resumeSession) {
            const hStore = hostSessionStore(hostRepoDir);
            const sessionPath = hStore.sessionFilePath(opts.resumeSession);
            if (!existsSync(sessionPath)) {
                throw new Error(`resumeSession "${opts.resumeSession}" not found: expected session file at ${sessionPath}`);
            }
        }
        const inner = Effect.gen(function* () {
            // 1. Resolve prompt
            const resolved = yield* resolvePrompt({ prompt, promptFile });
            const rawPrompt = resolved.text;
            const isInlinePrompt = resolved.source === "inline";
            // 2. Resolve env vars
            const resolvedEnv = yield* resolveEnv(hostRepoDir);
            const env = mergeProviderEnv({
                resolvedEnv,
                agentProviderEnv: provider.env,
                sandboxProviderEnv: sandboxProvider.env,
            });
            const effectiveEnv = { ...env, ...(opts.env ?? {}) };
            // 3. Prompt args substitution (skipped for inline prompts — passthrough)
            const userArgs = opts.promptArgs ?? {};
            let resolvedPrompt;
            if (isInlinePrompt) {
                yield* validateNoArgsWithInlinePrompt(userArgs);
                resolvedPrompt = rawPrompt;
            }
            else {
                yield* validateNoBuiltInArgOverride(userArgs);
                const effectiveArgs = {
                    SOURCE_BRANCH: worktreeInfo.branch,
                    TARGET_BRANCH: worktreeInfo.branch,
                    ...userArgs,
                };
                const builtInArgKeysSet = new Set(BUILT_IN_PROMPT_ARG_KEYS);
                resolvedPrompt = yield* substitutePromptArgs(rawPrompt, effectiveArgs, builtInArgKeysSet);
            }
            // 4. Start sandbox
            let handle;
            let sandboxRepoDir;
            if (sandboxProvider.tag === "isolated") {
                const startResult = yield* startSandbox({
                    provider: sandboxProvider,
                    hostRepoDir: worktreeInfo.path,
                    env: effectiveEnv,
                });
                handle = startResult.handle;
                sandboxRepoDir = startResult.worktreePath;
            }
            else {
                const gitPath = join(hostRepoDir, ".git");
                const gitMounts = yield* resolveGitMounts(gitPath);
                const startResult = yield* startSandbox({
                    provider: sandboxProvider,
                    hostRepoDir,
                    env: effectiveEnv,
                    worktreeOrRepoPath: worktreeInfo.path,
                    gitMounts,
                    repoDir: SANDBOX_REPO_DIR,
                });
                handle = startResult.handle;
                sandboxRepoDir = startResult.worktreePath;
            }
            const sandboxLayer = makeSandboxLayerFromHandle(handle);
            const applyToHost = sandboxProvider.tag === "isolated"
                ? () => syncOut(worktreeInfo.path, handle)
                : () => Effect.void;
            // 5. Resolve logging
            const resolvedLogging = opts.logging ?? {
                type: "file",
                path: join(hostRepoDir, ".sandcastle", "logs", buildLogFilename(worktreeInfo.branch, undefined, opts.name)),
            };
            const runDisplayLayer = resolvedLogging.type === "file"
                ? (() => {
                    printFileDisplayStartup({
                        logPath: resolvedLogging.path,
                        agentName: opts.name,
                        branch: worktreeInfo.branch,
                    });
                    return Layer.provide(FileDisplay.layer(resolvedLogging.path), NodeFileSystem.layer);
                })()
                : ClackDisplay.layer;
            // 6. Build a SandboxFactory that reuses the started sandbox
            const reuseFactoryLayer = Layer.succeed(SandboxFactory, {
                withSandbox: (makeEffect) => makeEffect({
                    hostWorktreePath: worktreeInfo.path,
                    sandboxRepoPath: sandboxRepoDir,
                    applyToHost,
                }).pipe(Effect.provide(sandboxLayer), Effect.map((value) => ({
                    value,
                    preservedWorktreePath: undefined,
                }))),
            });
            const agentStreamEmitterLayer = resolvedLogging.type === "file" && resolvedLogging.onAgentStreamEvent
                ? callbackAgentStreamEmitterLayer(resolvedLogging.onAgentStreamEvent)
                : noopAgentStreamEmitterLayer;
            const runLayer = Layer.mergeAll(reuseFactoryLayer, runDisplayLayer, defaultSessionPathsLayer, agentStreamEmitterLayer);
            // 7. Run orchestration
            const result = yield* Effect.gen(function* () {
                const display = yield* Display;
                yield* display.intro(opts.name ?? "sandcastle");
                return yield* orchestrate({
                    hostRepoDir,
                    iterations: maxIterations,
                    hooks,
                    prompt: resolvedPrompt,
                    branch: worktreeInfo.branch,
                    provider,
                    completionSignal: opts.completionSignal,
                    idleTimeoutSeconds: opts.idleTimeoutSeconds,
                    name: opts.name,
                    resumeSession: opts.resumeSession,
                    signal: opts.signal,
                    skipPromptExpansion: isInlinePrompt,
                });
            }).pipe(Effect.provide(runLayer), 
            // Always close sandbox handle
            Effect.ensuring(Effect.promise(() => handle.close().catch(() => { }))));
            return {
                iterations: result.iterations,
                completionSignal: result.completionSignal,
                stdout: result.stdout,
                commits: result.commits,
                branch: result.branch,
                logFilePath: resolvedLogging.type === "file" ? resolvedLogging.path : undefined,
            };
        });
        try {
            return await Effect.runPromise(inner.pipe(Effect.provide(ClackDisplay.layer), Effect.provide(NodeContext.layer), Effect.provide(NodeFileSystem.layer)));
        }
        catch (error) {
            // If the signal was aborted, surface its reason verbatim (no wrapping)
            opts.signal?.throwIfAborted();
            throw error;
        }
    };
    const worktreeCreateSandbox = async (opts) => {
        return createSandboxFromWorktree({
            branch: worktreeInfo.branch,
            worktreePath: worktreeInfo.path,
            hostRepoDir,
            sandbox: opts.sandbox,
            hooks: opts.hooks,
            copyToWorktree: opts.copyToWorktree,
            timeouts: opts.timeouts,
            _test: opts._test,
        });
    };
    return {
        branch: worktreeInfo.branch,
        worktreePath: worktreeInfo.path,
        run: worktreeRun,
        interactive: worktreeInteractive,
        createSandbox: worktreeCreateSandbox,
        close,
        async [Symbol.asyncDispose]() {
            await close();
        },
    };
};
