// @ai-hero/sandcastle/dist/Display.js
import * as clack from "@clack/prompts";
import { FileSystem } from "@effect/platform";
import { dirname } from "node:path";
import { Context, Effect, Layer, Ref } from "effect";
import { styleText } from "node:util";
export class Display extends Context.Tag("Display")() {
}
export const SilentDisplay = {
    layer: (ref) => Layer.succeed(Display, {
        intro: (title) => Ref.update(ref, (entries) => [
            ...entries,
            { _tag: "intro", title },
        ]),
        status: (message, severity) => Ref.update(ref, (entries) => [
            ...entries,
            { _tag: "status", message, severity },
        ]),
        spinner: (message, effect) => Effect.flatMap(Ref.update(ref, (entries) => [
            ...entries,
            { _tag: "spinner", message },
        ]), () => effect),
        summary: (title, rows) => Ref.update(ref, (entries) => [
            ...entries,
            { _tag: "summary", title, rows },
        ]),
        taskLog: (title, effect) => {
            const messages = [];
            return Effect.flatMap(effect((msg) => messages.push(msg)), (result) => Effect.map(Ref.update(ref, (entries) => [
                ...entries,
                {
                    _tag: "taskLog",
                    title,
                    messages: [...messages],
                },
            ]), () => result));
        },
        text: (message) => Ref.update(ref, (entries) => [
            ...entries,
            { _tag: "text", message },
        ]),
        toolCall: (name, formattedArgs) => Ref.update(ref, (entries) => [
            ...entries,
            { _tag: "toolCall", name, formattedArgs },
        ]),
    }),
};
export const FileDisplay = {
    layer: (filePath) => Layer.effect(Display, Effect.gen(function* () {
        const fs = yield* FileSystem.FileSystem;
        yield* fs
            .makeDirectory(dirname(filePath), { recursive: true })
            .pipe(Effect.orDie);
        const delimiter = `\n--- Run started: ${new Date().toISOString()} ---\n`;
        yield* fs
            .writeFileString(filePath, delimiter, { flag: "a" })
            .pipe(Effect.orDie);
        const appendToLog = (line) => fs
            .writeFileString(filePath, line + "\n", { flag: "a" })
            .pipe(Effect.ignore);
        return {
            intro: () => Effect.void,
            status: (message, _severity) => appendToLog(message.replace(/^\[[^\]]+\] /, "")),
            spinner: (message, effect) => Effect.gen(function* () {
                yield* appendToLog(`${message}...`);
                const start = Date.now();
                const result = yield* effect;
                const elapsed = ((Date.now() - start) / 1000).toFixed(1);
                yield* appendToLog(`${message} done (${elapsed}s)`);
                return result;
            }),
            summary: (title, rows) => {
                const lines = Object.entries(rows)
                    .map(([key, value]) => `  ${key}: ${value}`)
                    .join("\n");
                return appendToLog(`${title}\n${lines}`);
            },
            taskLog: (title, effect) => Effect.gen(function* () {
                yield* appendToLog(title);
                const start = Date.now();
                const messages = [];
                const result = yield* effect((msg) => {
                    messages.push(msg);
                });
                const elapsed = ((Date.now() - start) / 1000).toFixed(1);
                for (const msg of messages) {
                    yield* appendToLog(`  ${msg}`);
                }
                yield* appendToLog(`${title} done (${elapsed}s)`);
                return result;
            }),
            text: (message) => appendToLog(message),
            toolCall: (name, formattedArgs) => appendToLog(`${name}(${formattedArgs})`),
        };
    })),
};
const severityToClack = {
    info: clack.log.info,
    success: clack.log.success,
    warn: clack.log.warning,
    error: clack.log.error,
};
export const terminalStyle = {
    status: (message) => styleText("bold", message),
    summaryTitle: (title) => styleText("bold", title),
    summaryRow: (key, value) => `${styleText("bold", key)}: ${styleText("dim", value)}`,
    toolCall: (text) => styleText("dim", text),
};
export const ClackDisplay = {
    layer: Layer.succeed(Display, {
        intro: (title) => Effect.sync(() => clack.intro(styleText("inverse", ` ${title} `))),
        status: (message, severity) => Effect.sync(() => severityToClack[severity](terminalStyle.status(message))),
        spinner: (message, effect) => Effect.acquireUseRelease(Effect.sync(() => {
            const s = clack.spinner();
            s.start(message);
            return s;
        }), () => effect, (s, exit) => Effect.sync(() => {
            if (exit._tag === "Success") {
                s.stop(message);
            }
            else {
                s.stop(`${message} (failed)`);
            }
        })),
        summary: (title, rows) => Effect.sync(() => {
            const lines = Object.entries(rows)
                .map(([key, value]) => terminalStyle.summaryRow(key, value))
                .join("\n");
            clack.note(lines, terminalStyle.summaryTitle(title));
        }),
        taskLog: (title, effect) => Effect.acquireUseRelease(Effect.sync(() => clack.taskLog({ title })), (log) => effect((msg) => log.message(msg)), (log, exit) => Effect.sync(() => {
            if (exit._tag === "Success") {
                log.success(title, { showLog: true });
            }
            else {
                log.error(title, { showLog: true });
            }
        })),
        text: (message) => Effect.sync(() => clack.log.message(message)),
        toolCall: (name, formattedArgs) => Effect.sync(() => clack.log.step(terminalStyle.toolCall(`${name}(${formattedArgs})`))),
    }),
};
