// @ai-hero/sandcastle/dist/Orchestrator.js
import { Deferred, Effect } from "effect";
import { AgentStreamEmitter } from "./AgentStreamEmitter.js";
import { Display } from "./Display.js";
import { preprocessPrompt } from "./PromptPreprocessor.js";
import { AgentError, AgentIdleTimeoutError, SessionCaptureError, } from "./errors.js";
import { SandboxFactory } from "./SandboxFactory.js";
import { withSandboxLifecycle } from "./SandboxLifecycle.js";
import { TextDeltaBuffer } from "./TextDeltaBuffer.js";
import { hostSessionStore, sandboxSessionStore, transferSession, } from "./SessionStore.js";
import { SessionPaths } from "./SessionPaths.js";
const IDLE_WARNING_INTERVAL_MS = 60_000;
const invokeAgent = (sandbox, sandboxRepoDir, prompt, provider, idleTimeoutMs, onText, onToolCall, onIdleWarning, idleWarningIntervalMs = IDLE_WARNING_INTERVAL_MS, resumeSession, signal) => Effect.gen(function* () {
    let resultText = "";
    let sessionId;
    // Deferred that will be failed when the idle timer fires
    const timeoutSignal = yield* Deferred.make();
    let timeoutHandle = null;
    // Periodic idle warning state
    let warningHandle = null;
    let idleMinuteCounter = 0;
    const startWarningInterval = () => {
        if (warningHandle !== null)
            clearInterval(warningHandle);
        idleMinuteCounter = 0;
        warningHandle = setInterval(() => {
            idleMinuteCounter++;
            onIdleWarning(idleMinuteCounter);
        }, idleWarningIntervalMs);
    };
    const resetIdleTimer = () => {
        if (timeoutHandle !== null)
            clearTimeout(timeoutHandle);
        timeoutHandle = setTimeout(() => {
            Effect.runPromise(Deferred.fail(timeoutSignal, new AgentIdleTimeoutError({
                message: `Agent idle for ${idleTimeoutMs / 1000} seconds — no output received. Consider increasing the idle timeout with --idle-timeout.`,
                timeoutMs: idleTimeoutMs,
            }))).catch(() => { });
        }, idleTimeoutMs);
        // Reset warning interval on activity
        startWarningInterval();
    };
    // Deferred that will be resolved (as a defect) when the AbortSignal fires.
    // Uses Effect.die so the abort reason propagates as-is to run().
    const abortDeferred = yield* Deferred.make();
    let abortCleanup = null;
    if (signal) {
        if (signal.aborted) {
            return yield* Effect.die(signal.reason);
        }
        const onAbort = () => {
            Effect.runPromise(Deferred.die(abortDeferred, signal.reason)).catch(() => { });
        };
        signal.addEventListener("abort", onAbort, { once: true });
        abortCleanup = () => signal.removeEventListener("abort", onAbort);
    }
    resetIdleTimer();
    const execEffect = Effect.gen(function* () {
        const printCmd = provider.buildPrintCommand({
            prompt,
            dangerouslySkipPermissions: true,
            resumeSession,
        });
        const execResult = yield* sandbox.exec(printCmd.command, {
            onLine: (line) => {
                resetIdleTimer();
                for (const parsed of provider.parseStreamLine(line)) {
                    if (parsed.type === "text") {
                        onText(parsed.text);
                    }
                    else if (parsed.type === "result") {
                        resultText = parsed.result;
                    }
                    else if (parsed.type === "tool_call") {
                        onToolCall(parsed.name, parsed.args);
                    }
                    else if (parsed.type === "session_id") {
                        sessionId = parsed.sessionId;
                    }
                }
            },
            cwd: sandboxRepoDir,
            stdin: printCmd.stdin,
        });
        if (execResult.exitCode !== 0) {
            // Prefer stderr; fall back to resultText (from parsed stream events),
            // then to the tail of raw stdout (last 20 non-empty lines).
            let errorDetail = execResult.stderr;
            if (!errorDetail.trim()) {
                errorDetail = resultText;
            }
            if (!errorDetail.trim()) {
                const lines = execResult.stdout
                    .split("\n")
                    .filter((l) => l.trim());
                errorDetail = lines.slice(-20).join("\n");
            }
            return yield* Effect.fail(new AgentError({
                message: `${provider.name} exited with code ${execResult.exitCode}:\n${errorDetail}`,
            }));
        }
        return { result: resultText || execResult.stdout, sessionId };
    }).pipe(Effect.ensuring(Effect.sync(() => {
        if (timeoutHandle !== null) {
            clearTimeout(timeoutHandle);
            timeoutHandle = null;
        }
        if (warningHandle !== null) {
            clearInterval(warningHandle);
            warningHandle = null;
        }
    })));
    let raced = Effect.raceFirst(execEffect, Deferred.await(timeoutSignal));
    if (signal) {
        raced = Effect.raceFirst(raced, Deferred.await(abortDeferred));
    }
    return yield* raced.pipe(Effect.ensuring(Effect.sync(() => {
        abortCleanup?.();
    })));
});
const DEFAULT_COMPLETION_SIGNAL = "<promise>COMPLETE</promise>";
const DEFAULT_IDLE_TIMEOUT_SECONDS = 10 * 60; // 600 seconds
export const orchestrate = (options) => {
    const idleTimeoutMs = (options.idleTimeoutSeconds ?? DEFAULT_IDLE_TIMEOUT_SECONDS) * 1000;
    return Effect.gen(function* () {
        const factory = yield* SandboxFactory;
        const display = yield* Display;
        const streamEmitter = yield* AgentStreamEmitter;
        const { hostProjectsDir, sandboxProjectsDir } = yield* SessionPaths;
        const { hostRepoDir, iterations, hooks, prompt, branch, provider } = options;
        let completionSignals;
        if (options.completionSignal === undefined) {
            completionSignals = [DEFAULT_COMPLETION_SIGNAL];
        }
        else if (Array.isArray(options.completionSignal)) {
            completionSignals = options.completionSignal;
        }
        else {
            completionSignals = [options.completionSignal];
        }
        const label = (msg) => options.name ? `[${options.name}] ${msg}` : msg;
        const allCommits = [];
        const allIterations = [];
        let allStdout = "";
        let resolvedBranch = "";
        let iterationPreservedPath;
        // Helper: check abort signal and bail via defect so run() can
        // re-throw the signal's reason verbatim (no Sandcastle wrapping).
        const checkAbort = () => options.signal?.aborted ? Effect.die(options.signal.reason) : Effect.void;
        for (let i = 1; i <= iterations; i++) {
            yield* checkAbort();
            yield* display.status(label(`Iteration ${i}/${iterations}`), "info");
            const sandboxResult = yield* factory.withSandbox(({ hostWorktreePath, sandboxRepoPath, applyToHost, bindMountHandle }) => withSandboxLifecycle({
                hostRepoDir,
                sandboxRepoDir: sandboxRepoPath,
                hooks,
                branch,
                hostWorktreePath,
                applyToHost,
                signal: options.signal,
            }, (ctx) => Effect.gen(function* () {
                // Resume session: transfer JSONL from host to sandbox before iteration 1
                const iterationResumeSession = i === 1 ? options.resumeSession : undefined;
                if (iterationResumeSession && bindMountHandle) {
                    yield* display.status(label("Resuming session"), "info");
                    const sbStore = sandboxSessionStore(ctx.sandboxRepoDir, bindMountHandle, sandboxProjectsDir);
                    const hStore = hostSessionStore(hostRepoDir, hostProjectsDir);
                    yield* Effect.tryPromise({
                        try: () => transferSession(hStore, sbStore, iterationResumeSession),
                        catch: (e) => new SessionCaptureError({
                            message: `Session resume failed: ${e instanceof Error ? e.message : String(e)}`,
                            sessionId: iterationResumeSession,
                        }),
                    });
                }
                // Preprocess prompt (run !`command` expressions inside sandbox).
                // Inline prompts pass through literally — skip expansion.
                const fullPrompt = options.skipPromptExpansion
                    ? prompt
                    : yield* preprocessPrompt(prompt, ctx.sandbox, ctx.sandboxRepoDir);
                yield* display.status(label("Agent started"), "success");
                // Invoke the agent — buffer text deltas so Pi's single-token
                // chunks are displayed as readable multi-word lines.
                const textBuffer = new TextDeltaBuffer((chunk) => {
                    Effect.runPromise(display.text(chunk));
                    Effect.runPromise(streamEmitter.emit({
                        type: "text",
                        message: chunk,
                        iteration: i,
                        timestamp: new Date(),
                    }));
                });
                const onText = (text) => {
                    textBuffer.write(text);
                };
                const onToolCall = (name, formattedArgs) => {
                    textBuffer.flush();
                    Effect.runPromise(display.toolCall(name, formattedArgs));
                    Effect.runPromise(streamEmitter.emit({
                        type: "toolCall",
                        name,
                        formattedArgs,
                        iteration: i,
                        timestamp: new Date(),
                    }));
                };
                const onIdleWarning = (minutes) => {
                    const msg = minutes === 1
                        ? "Agent idle for 1 minute"
                        : `Agent idle for ${minutes} minutes`;
                    Effect.runPromise(display.status(label(msg), "warn"));
                };
                const { result: agentOutput, sessionId } = yield* invokeAgent(ctx.sandbox, ctx.sandboxRepoDir, fullPrompt, provider, idleTimeoutMs, onText, onToolCall, onIdleWarning, options._idleWarningIntervalMs, iterationResumeSession, options.signal);
                // Flush any remaining buffered text deltas
                textBuffer.dispose();
                yield* display.status(label("Agent stopped"), "info");
                // Capture session while sandbox is still alive
                let sessionFilePath;
                let usage;
                if (provider.captureSessions && sessionId && bindMountHandle) {
                    yield* display.status(label("Capturing session"), "info");
                    const sbStore = sandboxSessionStore(ctx.sandboxRepoDir, bindMountHandle, sandboxProjectsDir);
                    const hStore = hostSessionStore(hostRepoDir, hostProjectsDir);
                    yield* Effect.tryPromise({
                        try: () => transferSession(sbStore, hStore, sessionId),
                        catch: (e) => new SessionCaptureError({
                            message: `Session capture failed: ${e instanceof Error ? e.message : String(e)}`,
                            sessionId,
                        }),
                    });
                    sessionFilePath = hStore.sessionFilePath(sessionId);
                    // Parse token usage from the captured session JSONL
                    if (provider.parseSessionUsage) {
                        const content = yield* Effect.promise(() => hStore
                            .readSession(sessionId)
                            .catch(() => undefined));
                        if (content) {
                            usage = provider.parseSessionUsage(content);
                        }
                    }
                }
                // Check completion signal
                const matchedSignal = completionSignals.find((sig) => agentOutput.includes(sig));
                return {
                    completionSignal: matchedSignal,
                    stdout: agentOutput,
                    sessionId,
                    sessionFilePath,
                    usage,
                };
            })));
            const lifecycleResult = sandboxResult.value;
            iterationPreservedPath = sandboxResult.preservedWorktreePath;
            allCommits.push(...lifecycleResult.commits);
            allStdout += lifecycleResult.result.stdout;
            resolvedBranch = lifecycleResult.branch;
            allIterations.push({
                sessionId: lifecycleResult.result.sessionId,
                sessionFilePath: lifecycleResult.result.sessionFilePath,
                usage: lifecycleResult.result.usage,
            });
            if (lifecycleResult.result.completionSignal !== undefined) {
                yield* display.status(label(`Agent signaled completion after ${i} iteration(s).`), "success");
                return {
                    iterations: allIterations,
                    completionSignal: lifecycleResult.result.completionSignal,
                    stdout: allStdout,
                    commits: allCommits,
                    branch: resolvedBranch,
                    preservedWorktreePath: iterationPreservedPath,
                };
            }
        }
        yield* display.status(label(`Reached max iterations (${iterations}).`), "info");
        return {
            iterations: allIterations,
            completionSignal: undefined,
            stdout: allStdout,
            commits: allCommits,
            branch: resolvedBranch,
            preservedWorktreePath: iterationPreservedPath,
        };
    });
};
