// @ai-hero/sandcastle/dist/PromptArgumentSubstitution.js
import { Effect } from "effect";
import { Display } from "./Display.js";
import { PromptError } from "./errors.js";
import { SHELL_BLOCK_MARKER } from "./PromptPreprocessor.js";
const SHELL_BLOCK_PATTERN = /!`([^`]+)`/g;
/**
 * Prompt argument keys that Sandcastle injects automatically.
 * These cannot be overridden via `promptArgs`.
 */
export const BUILT_IN_PROMPT_ARG_KEYS = [
    "SOURCE_BRANCH",
    "TARGET_BRANCH",
];
const PLACEHOLDER_PATTERN = /\{\{\s*([A-Za-z_][A-Za-z0-9_]*)\s*\}\}/g;
/**
 * Validates that the user has not provided any built-in prompt argument keys.
 * Fails with a PromptError if any built-in key is found in `args`.
 */
/**
 * Fails if the user passed any `promptArgs` alongside an inline prompt.
 * Inline prompts are delivered to the agent verbatim, so `promptArgs` would
 * silently do nothing. An empty object is treated as absent.
 */
export const validateNoArgsWithInlinePrompt = (args) => {
    if (Object.keys(args).length === 0)
        return Effect.void;
    return Effect.fail(new PromptError({
        message: 'promptArgs is only supported with promptFile. Inline prompts (prompt: "...") are passed to the agent as-is — interpolate values directly in JavaScript, or switch to promptFile to use {{KEY}} substitution.',
    }));
};
export const validateNoBuiltInArgOverride = (args) => {
    for (const key of BUILT_IN_PROMPT_ARG_KEYS) {
        if (key in args) {
            return Effect.fail(new PromptError({
                message: `"${key}" is a built-in prompt argument and cannot be overridden via promptArgs`,
            }));
        }
    }
    return Effect.void;
};
/**
 * Scans a prompt for `{{KEY}}` placeholders and returns the keys that are
 * missing from `providedArgs`, excluding built-in keys.
 */
export const findMissingPromptArgKeys = (prompt, providedArgs) => {
    const matches = [...prompt.matchAll(PLACEHOLDER_PATTERN)];
    const builtInSet = new Set(BUILT_IN_PROMPT_ARG_KEYS);
    const seen = new Set();
    const missing = [];
    for (const m of matches) {
        const key = m[1];
        if (seen.has(key))
            continue;
        seen.add(key);
        if (builtInSet.has(key))
            continue;
        if (key in providedArgs)
            continue;
        missing.push(key);
    }
    return missing;
};
export const substitutePromptArgs = (prompt, args, silentKeys) => {
    // Mark shell blocks written in the raw template so the preprocessor can
    // distinguish them from `!`...`` patterns that arrive via arg substitution.
    // Strip any markers already present in raw input so they can't be forged.
    const markedPrompt = prompt
        .replaceAll(SHELL_BLOCK_MARKER, "")
        .replace(SHELL_BLOCK_PATTERN, `!${SHELL_BLOCK_MARKER}\`$1\``);
    const sanitizedArgs = Object.fromEntries(Object.entries(args).map(([key, value]) => [
        key,
        typeof value === "string"
            ? value.replaceAll(SHELL_BLOCK_MARKER, "")
            : value,
    ]));
    const matches = [...markedPrompt.matchAll(PLACEHOLDER_PATTERN)];
    if (matches.length === 0 && Object.keys(sanitizedArgs).length === 0) {
        return Effect.succeed(markedPrompt);
    }
    return Effect.gen(function* () {
        const display = yield* Display;
        // Collect all keys referenced in the prompt
        const referencedKeys = new Set(matches.map((m) => m[1]));
        // Check for missing keys (placeholder in prompt but no matching arg)
        for (const key of referencedKeys) {
            if (!(key in sanitizedArgs)) {
                return yield* Effect.fail(new PromptError({
                    message: `Prompt argument "{{${key}}}" has no matching value in promptArgs`,
                }));
            }
        }
        // Warn about unused keys (arg provided but no matching placeholder)
        // Skip keys listed in silentKeys (e.g. built-in args)
        for (const key of Object.keys(sanitizedArgs)) {
            if (!referencedKeys.has(key) && !silentKeys?.has(key)) {
                yield* display.status(`Prompt argument "${key}" was provided but not referenced in the prompt`, "warn");
            }
        }
        // Replace all placeholders with their values
        const result = markedPrompt.replace(PLACEHOLDER_PATTERN, (_match, key) => sanitizedArgs[key].toString());
        return result;
    });
};
