// @ai-hero/sandcastle/dist/run.js
import { NodeContext, NodeFileSystem } from "@effect/platform-node";
import { existsSync } from "node:fs";
import path, { join } from "node:path";
import { styleText } from "node:util";
import { Effect, Layer } from "effect";
import { resolveCwd } from "./resolveCwd.js";
import { ClackDisplay, Display, FileDisplay, } from "./Display.js";
import { orchestrate, } from "./Orchestrator.js";
import { resolvePrompt } from "./PromptResolver.js";
import { WorktreeDockerSandboxFactory, SandboxConfig, } from "./SandboxFactory.js";
import { resolveEnv } from "./EnvResolver.js";
import { formatErrorMessage } from "./ErrorHandler.js";
import { callbackAgentStreamEmitterLayer, noopAgentStreamEmitterLayer, } from "./AgentStreamEmitter.js";
import { mergeProviderEnv } from "./mergeProviderEnv.js";
import { hostSessionStore } from "./SessionStore.js";
import { defaultSessionPathsLayer } from "./SessionPaths.js";
import { generateTempBranchName, getCurrentBranch } from "./WorktreeManager.js";
import { substitutePromptArgs, validateNoArgsWithInlinePrompt, validateNoBuiltInArgOverride, BUILT_IN_PROMPT_ARG_KEYS, } from "./PromptArgumentSubstitution.js";
import { extractStructuredOutput } from "./extractStructuredOutput.js";
/** Default maximum number of iterations for a run. */
export const DEFAULT_MAX_ITERATIONS = 1;
/** Replace characters that are invalid or problematic in file paths with dashes. */
export const sanitizeBranchForFilename = (branch) => branch.replace(/[/\\:*?"<>|]/g, "-");
/**
 * Print the startup message to the terminal when using file-based logging.
 * Uses styleText for lightweight bold/dim styling — does not use Clack.
 */
export const printFileDisplayStartup = (options) => {
    const name = options.agentName ?? "Agent";
    const label = styleText("bold", `[${name}]`);
    const branchPart = options.branch ? ` on branch ${options.branch}` : "";
    const hostRepoDir = options.hostRepoDir ?? process.cwd();
    const displayLogPath = hostRepoDir === process.cwd()
        ? path.relative(process.cwd(), options.logPath)
        : options.logPath;
    console.log(`${label} Started${branchPart}`);
    console.log(styleText("dim", `  tail -f ${displayLogPath}`));
};
/**
 * Build the log filename for a run.
 * When a targetBranch is provided (temp branch mode), prefixes the filename
 * with the sanitized target branch name so developers can identify which
 * branch the run was targeting: `<targetBranch>-<resolvedBranch>.log`
 * When no targetBranch, uses just the resolved branch: `<resolvedBranch>.log`
 * When a name is provided, appends it to avoid collisions in multi-agent workflows.
 */
export const buildLogFilename = (resolvedBranch, targetBranch, name) => {
    const sanitized = sanitizeBranchForFilename(resolvedBranch);
    const nameSuffix = name
        ? `-${name.toLowerCase().replace(/[^a-z0-9_.-]/g, "-")}`
        : "";
    if (targetBranch) {
        return `${sanitizeBranchForFilename(targetBranch)}-${sanitized}${nameSuffix}.log`;
    }
    return `${sanitized}${nameSuffix}.log`;
};
/**
 * Build the summary rows for a run, used in both terminal mode and
 * log-to-file mode. When a custom name is provided it appears as the
 * Agent value instead of the internal provider name.
 */
export const buildRunSummaryRows = (options) => ({
    Agent: options.name ?? options.agentName,
    Sandbox: options.sandboxName,
    "Max iterations": String(options.maxIterations),
    Branch: options.branch,
});
/**
 * Build the completion status message for a run, used in both terminal mode
 * and log-to-file mode to record the final outcome.
 */
export const buildCompletionMessage = (completionSignal, iterationsRun) => {
    if (completionSignal !== undefined) {
        return {
            message: `Run complete: agent finished after ${iterationsRun} iteration(s).`,
            severity: "success",
        };
    }
    return {
        message: `Run complete: reached ${iterationsRun} iteration(s) without completion signal.`,
        severity: "warn",
    };
};
/**
 * Format the context window size from an iteration's usage data.
 * Returns a string like "103k" representing the total input-side tokens
 * (inputTokens + cacheCreationInputTokens + cacheReadInputTokens)
 * rounded up to the nearest 1000.
 */
export const formatContextWindowSize = (usage) => {
    const total = usage.inputTokens +
        usage.cacheCreationInputTokens +
        usage.cacheReadInputTokens;
    return `${Math.ceil(total / 1000)}k`;
};
/**
 * Build "Context window: NNNk" lines for iterations that have usage data.
 * Returns an empty array when no iterations carry usage.
 */
export const buildContextWindowLines = (iterations) => iterations
    .filter((it) => it.usage !== undefined)
    .map((it) => `Context window: ${formatContextWindowSize(it.usage)}`);
export async function run(options) {
    // If signal is already aborted, reject immediately without any setup
    options.signal?.throwIfAborted();
    const { prompt, promptFile, maxIterations = DEFAULT_MAX_ITERATIONS, hooks, agent: provider, } = options;
    // Derive branch strategy: explicit option > default based on provider tag
    const branchStrategy = options.branchStrategy ??
        (options.sandbox.tag === "isolated"
            ? { type: "merge-to-head" }
            : { type: "head" });
    const effectiveBranchType = branchStrategy.type;
    // Validate: head strategy is not supported with isolated providers
    if (effectiveBranchType === "head" && options.sandbox.tag === "isolated") {
        throw new Error("head branch strategy is not supported with isolated providers");
    }
    // Validate: copyToWorktree is incompatible with head strategy
    if (effectiveBranchType === "head" &&
        options.copyToWorktree &&
        options.copyToWorktree.length > 0) {
        throw new Error("copyToWorktree is not supported with head branch strategy. " +
            "In head mode the host working directory is bind-mounted directly.");
    }
    // Validate: resumeSession + maxIterations > 1 is not allowed
    if (options.resumeSession && maxIterations > 1) {
        throw new Error("resumeSession cannot be combined with maxIterations > 1. " +
            "Resume applies to iteration 1 only; multi-iteration resume semantics are not supported.");
    }
    // Validate: output requires maxIterations === 1
    if (options.output && maxIterations !== 1) {
        throw new Error("output requires maxIterations to be 1. " +
            "Structured output is only supported for single-iteration runs.");
    }
    // Extract explicit branch when in branch mode
    const branch = branchStrategy.type === "branch" ? branchStrategy.branch : undefined;
    const hostRepoDir = await Effect.runPromise(resolveCwd(options.cwd).pipe(Effect.provide(NodeContext.layer)));
    // Validate: resumeSession file must exist on the host
    if (options.resumeSession) {
        const hStore = hostSessionStore(hostRepoDir);
        const sessionPath = hStore.sessionFilePath(options.resumeSession);
        if (!existsSync(sessionPath)) {
            throw new Error(`resumeSession "${options.resumeSession}" not found: expected session file at ${sessionPath}`);
        }
    }
    // Resolve prompt
    const resolved = await Effect.runPromise(resolvePrompt({ prompt, promptFile }).pipe(Effect.provide(NodeContext.layer)));
    const rawPrompt = resolved.text;
    const isInlinePrompt = resolved.source === "inline";
    // Validate: output tag must appear in the resolved prompt
    if (options.output) {
        const openTag = `<${options.output.tag}>`;
        if (!rawPrompt.includes(openTag)) {
            throw new Error(`output tag <${options.output.tag}> not found in the resolved prompt. ` +
                "The caller must instruct the agent to emit the configured tag.");
        }
    }
    const agentName = provider.name;
    // Resolve env vars and merge with provider env
    const resolvedEnv = await Effect.runPromise(resolveEnv(hostRepoDir).pipe(Effect.provide(NodeContext.layer)));
    const env = mergeProviderEnv({
        resolvedEnv,
        agentProviderEnv: provider.env,
        sandboxProviderEnv: options.sandbox.env,
    });
    // Always capture the host's current branch for the TARGET_BRANCH built-in
    // prompt argument. When using a temp branch, it also prefixes the log filename.
    const currentHostBranch = await Effect.runPromise(getCurrentBranch(hostRepoDir));
    // When in merge-to-head mode, generate a temporary branch name.
    // In head mode, use the host's current branch directly (no worktree).
    const resolvedBranch = effectiveBranchType === "head"
        ? currentHostBranch
        : (branch ?? generateTempBranchName(options.name));
    // When using a temp branch, prefix the log filename with the target branch
    // (the host's current branch) so developers can tell which branch was targeted.
    const targetBranch = effectiveBranchType === "merge-to-head" ? currentHostBranch : undefined;
    // Resolve logging option
    const resolvedLogging = options.logging ?? {
        type: "file",
        path: join(hostRepoDir, ".sandcastle", "logs", buildLogFilename(resolvedBranch, targetBranch, options.name)),
    };
    const displayLayer = resolvedLogging.type === "file"
        ? (() => {
            printFileDisplayStartup({
                logPath: resolvedLogging.path,
                agentName: options.name,
                branch: resolvedBranch,
                hostRepoDir,
            });
            return Layer.provide(FileDisplay.layer(resolvedLogging.path), NodeFileSystem.layer);
        })()
        : ClackDisplay.layer;
    const factoryLayer = Layer.provide(WorktreeDockerSandboxFactory.layer, Layer.mergeAll(Layer.succeed(SandboxConfig, {
        env,
        hostRepoDir,
        copyToWorktree: options.copyToWorktree,
        name: options.name,
        sandboxProvider: options.sandbox,
        branchStrategy,
        hooks,
        signal: options.signal,
        timeouts: options.timeouts,
    }), NodeFileSystem.layer, displayLayer));
    const agentStreamEmitterLayer = resolvedLogging.type === "file" && resolvedLogging.onAgentStreamEvent
        ? callbackAgentStreamEmitterLayer(resolvedLogging.onAgentStreamEvent)
        : noopAgentStreamEmitterLayer;
    const runLayer = Layer.mergeAll(factoryLayer, displayLayer, defaultSessionPathsLayer, agentStreamEmitterLayer);
    const baseEffect = Effect.gen(function* () {
        const d = yield* Display;
        yield* d.intro(options.name ?? "sandcastle");
        const rows = buildRunSummaryRows({
            name: options.name,
            agentName,
            sandboxName: options.sandbox.name,
            maxIterations,
            branch: resolvedBranch,
        });
        yield* d.summary("Sandcastle Run", rows);
        const userArgs = options.promptArgs ?? {};
        // Inline prompts pass through to the agent literally — no substitution,
        // no built-in arg injection. Guard against silently ignoring promptArgs.
        let resolvedPrompt;
        if (isInlinePrompt) {
            yield* validateNoArgsWithInlinePrompt(userArgs);
            resolvedPrompt = rawPrompt;
        }
        else {
            yield* validateNoBuiltInArgOverride(userArgs);
            const effectiveArgs = {
                SOURCE_BRANCH: resolvedBranch,
                TARGET_BRANCH: currentHostBranch,
                ...userArgs,
            };
            const builtInArgKeysSet = new Set(BUILT_IN_PROMPT_ARG_KEYS);
            resolvedPrompt = yield* substitutePromptArgs(rawPrompt, effectiveArgs, builtInArgKeysSet);
        }
        // In head mode, pass the host branch so SandboxLifecycle skips the merge step.
        // In merge-to-head mode, branch is undefined (triggers merge). In branch mode, it's the explicit branch.
        const orchestrateBranch = effectiveBranchType === "head" ? currentHostBranch : branch;
        const orchestrateResult = yield* orchestrate({
            hostRepoDir,
            iterations: maxIterations,
            hooks,
            prompt: resolvedPrompt,
            branch: orchestrateBranch,
            provider,
            completionSignal: options.completionSignal,
            idleTimeoutSeconds: options.idleTimeoutSeconds,
            name: options.name,
            resumeSession: options.resumeSession,
            signal: options.signal,
            skipPromptExpansion: isInlinePrompt,
        });
        const completion = buildCompletionMessage(orchestrateResult.completionSignal, orchestrateResult.iterations.length);
        yield* d.status(completion.message, completion.severity);
        for (const line of buildContextWindowLines(orchestrateResult.iterations)) {
            yield* d.text(line);
        }
        return orchestrateResult;
    });
    // In file-logging mode, write errors to the log before they propagate.
    // In stdout mode (ClackDisplay), errors are already shown by withFriendlyErrors
    // in main.ts, so we skip to avoid duplicate terminal output.
    const withErrorLog = resolvedLogging.type === "file"
        ? baseEffect.pipe(Effect.tapError((error) => Effect.gen(function* () {
            const d = yield* Display;
            yield* d.status(formatErrorMessage(error), "error");
        })))
        : baseEffect;
    let result;
    try {
        result = await Effect.runPromise(withErrorLog.pipe(Effect.provide(runLayer)));
    }
    catch (error) {
        // If the signal was aborted, surface its reason verbatim (no wrapping)
        options.signal?.throwIfAborted();
        throw error;
    }
    const baseResult = {
        ...result,
        logFilePath: resolvedLogging.type === "file" ? resolvedLogging.path : undefined,
    };
    // Extract structured output after the iteration completes (separate pass from completion signal)
    if (options.output) {
        const output = await extractStructuredOutput(baseResult.stdout, options.output, {
            commits: baseResult.commits,
            branch: baseResult.branch,
            preservedWorktreePath: baseResult.preservedWorktreePath,
        });
        return { ...baseResult, output };
    }
    return baseResult;
}
