// @ai-hero/sandcastle/dist/sandboxes/daytona.js
/**
 * Daytona isolated sandbox provider.
 *
 * Creates ephemeral Daytona sandboxes via `@daytona/sdk`.
 * Requires `@daytona/sdk` as a peer dependency.
 */
import { readdir, stat } from "node:fs/promises";
import { join, relative } from "node:path";
import { createIsolatedSandboxProvider, } from "../SandboxProvider.js";
/**
 * Create a Daytona isolated sandbox provider.
 *
 * Sandboxes are ephemeral — each `create()` call spins up a new Daytona
 * sandbox and `close()` destroys it.
 *
 * @example
 * ```ts
 * import { daytona } from "@ai-hero/sandcastle/sandboxes/daytona";
 *
 * const provider = daytona({ apiKey: "dyt_my_key" });
 * ```
 */
export const daytona = (options) => createIsolatedSandboxProvider({
    name: "daytona",
    env: options?.env,
    create: async () => {
        const { Daytona } = (await import("@daytona/sdk"));
        const config = {};
        if (options?.apiKey)
            config.apiKey = options.apiKey;
        if (options?.apiUrl)
            config.apiUrl = options.apiUrl;
        if (options?.target)
            config.target = options.target;
        const client = new Daytona(config);
        const sandbox = await client.create(options?.create);
        const worktreePath = (await sandbox.getWorkDir()) ??
            (await sandbox.getUserHomeDir()) ??
            "/home/daytona";
        return {
            worktreePath,
            exec: async (command, opts) => {
                const effectiveCommand = opts?.sudo ? `sudo ${command}` : command;
                if (opts?.onLine) {
                    const onLine = opts.onLine;
                    const sessionId = `sandcastle-${crypto.randomUUID()}`;
                    await sandbox.process.createSession(sessionId);
                    try {
                        const execResponse = await sandbox.process.executeSessionCommand(sessionId, {
                            command: `cd ${opts?.cwd ?? worktreePath} && ${effectiveCommand}`,
                            async: true,
                        });
                        const cmdId = execResponse.cmdId;
                        const stdoutLines = [];
                        const stderrChunks = [];
                        let partial = "";
                        await sandbox.process.getSessionCommandLogs(sessionId, cmdId, (chunk) => {
                            const text = partial + chunk;
                            const lines = text.split("\n");
                            partial = lines.pop() ?? "";
                            for (const line of lines) {
                                stdoutLines.push(line);
                                onLine(line);
                            }
                        }, (chunk) => {
                            stderrChunks.push(chunk);
                        });
                        if (partial) {
                            stdoutLines.push(partial);
                            onLine(partial);
                        }
                        const cmdInfo = await sandbox.process.getSessionCommand(sessionId, cmdId);
                        return {
                            stdout: stdoutLines.join("\n"),
                            stderr: stderrChunks.join(""),
                            exitCode: cmdInfo.exitCode ?? 0,
                        };
                    }
                    finally {
                        await sandbox.process.deleteSession(sessionId).catch(() => { });
                    }
                }
                const response = await sandbox.process.executeCommand(effectiveCommand, opts?.cwd ?? worktreePath);
                return {
                    stdout: response.result,
                    stderr: "",
                    exitCode: response.exitCode,
                };
            },
            copyIn: async (hostPath, sandboxPath) => {
                const info = await stat(hostPath);
                if (info.isDirectory()) {
                    const walk = async (dir) => {
                        const entries = await readdir(dir, { withFileTypes: true });
                        const files = [];
                        for (const entry of entries) {
                            const full = join(dir, entry.name);
                            if (entry.isDirectory()) {
                                files.push(...(await walk(full)));
                            }
                            else {
                                files.push(full);
                            }
                        }
                        return files;
                    };
                    const files = await walk(hostPath);
                    for (const file of files) {
                        const rel = relative(hostPath, file);
                        await sandbox.fs.uploadFile(file, join(sandboxPath, rel));
                    }
                }
                else {
                    await sandbox.fs.uploadFile(hostPath, sandboxPath);
                }
            },
            copyFileOut: async (sandboxPath, hostPath) => {
                await sandbox.fs.downloadFile(sandboxPath, hostPath);
            },
            close: async () => {
                await client.delete(sandbox);
            },
        };
    },
});
