// @ai-hero/sandcastle/dist/sandboxes/docker.js
/**
 * Docker sandbox provider — wraps DockerLifecycle into a SandboxProvider.
 *
 * Usage:
 *   import { docker } from "sandcastle/sandboxes/docker";
 *   await run({ agent: claudeCode("claude-opus-4-6"), sandbox: docker() });
 */
import { execFile, execFileSync, spawn, } from "node:child_process";
import { randomUUID } from "node:crypto";
import { createInterface } from "node:readline";
import { Effect } from "effect";
import { startContainer, removeContainer } from "../DockerLifecycle.js";
import { createBindMountSandboxProvider, } from "../SandboxProvider.js";
import { defaultImageName, resolveUserMounts, processFileMountParents, } from "../mountUtils.js";
/**
 * Create a Docker sandbox provider.
 *
 * The returned provider creates Docker containers with bind-mounts
 * for the worktree and git directories.
 */
export const docker = (options) => {
    const configuredImageName = options?.imageName;
    const selinuxLabel = options?.selinuxLabel ?? "z";
    const sandboxHomedir = "/home/agent";
    const userMounts = options?.mounts
        ? resolveUserMounts(options.mounts, sandboxHomedir)
        : [];
    // Validate file mounts and collect parent dirs to create at container start.
    // Throws at construction time if any file mount parent is outside sandboxHomedir.
    const parentDirsToCreate = processFileMountParents(userMounts, sandboxHomedir);
    return createBindMountSandboxProvider({
        name: "docker",
        env: options?.env,
        sandboxHomedir,
        create: async (createOptions) => {
            const containerName = `sandcastle-${randomUUID()}`;
            const worktreePath = createOptions.mounts.find((m) => m.hostPath === createOptions.worktreePath)?.sandboxPath ?? "/home/agent/workspace";
            // Build volume mount list (internal mounts + user-provided mounts)
            const allMounts = [...createOptions.mounts, ...userMounts];
            const volumeMounts = allMounts.map((m) => ({
                hostPath: m.hostPath,
                sandboxPath: m.sandboxPath,
                readonly: m.readonly,
            }));
            // Resolve image name
            const imageName = configuredImageName ?? defaultImageName(createOptions.hostRepoPath);
            const containerUid = options?.containerUid ?? process.getuid?.() ?? 1000;
            const containerGid = options?.containerGid ?? process.getgid?.() ?? 1000;
            // Pre-flight: verify image exists and UID matches
            await checkImageUid(imageName, containerUid);
            // Start container
            await Effect.runPromise(startContainer(containerName, imageName, {
                ...createOptions.env,
                HOME: "/home/agent",
            }, {
                volumeMounts,
                workdir: worktreePath,
                user: `${containerUid}:${containerGid}`,
                network: options?.network,
                selinuxLabel,
            }));
            // Create parent directories for file mounts and chown to the container user
            for (const dir of parentDirsToCreate) {
                await new Promise((resolve, reject) => {
                    execFile("docker", [
                        "exec",
                        "--user",
                        "0:0",
                        containerName,
                        "sh",
                        "-c",
                        `mkdir -p "$1" && chown "$2" "$1"`,
                        "sh",
                        dir,
                        `${containerUid}:${containerGid}`,
                    ], (error) => {
                        if (error) {
                            reject(new Error(`Failed to create parent directory '${dir}' in container: ${error.message}`));
                        }
                        else {
                            resolve();
                        }
                    });
                });
            }
            // Set up signal handlers for cleanup
            const onExit = () => {
                try {
                    execFileSync("docker", ["rm", "-f", containerName], {
                        stdio: "ignore",
                    });
                }
                catch {
                    /* best-effort */
                }
            };
            const onSignal = () => {
                onExit();
                process.exit(1);
            };
            process.on("exit", onExit);
            process.on("SIGINT", onSignal);
            process.on("SIGTERM", onSignal);
            const handle = {
                worktreePath,
                exec: (command, opts) => {
                    const effectiveCommand = opts?.sudo ? `sudo ${command}` : command;
                    const args = ["exec"];
                    if (opts?.stdin !== undefined)
                        args.push("-i");
                    if (opts?.cwd)
                        args.push("-w", opts.cwd);
                    args.push(containerName, "sh", "-c", effectiveCommand);
                    return new Promise((resolve, reject) => {
                        const proc = spawn("docker", args, {
                            stdio: [
                                opts?.stdin !== undefined ? "pipe" : "ignore",
                                "pipe",
                                "pipe",
                            ],
                        });
                        if (opts?.stdin !== undefined) {
                            proc.stdin.write(opts.stdin);
                            proc.stdin.end();
                        }
                        const stdoutChunks = [];
                        const stderrChunks = [];
                        if (opts?.onLine) {
                            const onLine = opts.onLine;
                            const rl = createInterface({ input: proc.stdout });
                            rl.on("line", (line) => {
                                stdoutChunks.push(line);
                                onLine(line);
                            });
                        }
                        else {
                            proc.stdout.on("data", (chunk) => {
                                stdoutChunks.push(chunk.toString());
                            });
                        }
                        proc.stderr.on("data", (chunk) => {
                            stderrChunks.push(chunk.toString());
                        });
                        proc.on("error", (error) => {
                            reject(new Error(`docker exec failed: ${error.message}`));
                        });
                        proc.on("close", (code) => {
                            resolve({
                                stdout: stdoutChunks.join(opts?.onLine ? "\n" : ""),
                                stderr: stderrChunks.join(""),
                                exitCode: code ?? 0,
                            });
                        });
                    });
                },
                interactiveExec: (args, opts) => {
                    return new Promise((resolve, reject) => {
                        const dockerArgs = ["exec"];
                        // Allocate a pseudo-terminal when stdin looks like a TTY
                        if ("isTTY" in opts.stdin &&
                            opts.stdin.isTTY) {
                            dockerArgs.push("-it");
                        }
                        else {
                            dockerArgs.push("-i");
                        }
                        if (opts.cwd)
                            dockerArgs.push("-w", opts.cwd);
                        dockerArgs.push(containerName, ...args);
                        const proc = spawn("docker", dockerArgs, {
                            stdio: [opts.stdin, opts.stdout, opts.stderr],
                        });
                        proc.on("error", (error) => {
                            reject(new Error(`docker exec failed: ${error.message}`));
                        });
                        proc.on("close", (code) => {
                            resolve({ exitCode: code ?? 0 });
                        });
                    });
                },
                copyFileIn: (hostPath, sandboxPath) => new Promise((resolve, reject) => {
                    execFile("docker", ["cp", hostPath, `${containerName}:${sandboxPath}`], (error) => {
                        if (error) {
                            reject(new Error(`docker cp (in) failed: ${error.message}`));
                        }
                        else {
                            resolve();
                        }
                    });
                }),
                copyFileOut: (sandboxPath, hostPath) => new Promise((resolve, reject) => {
                    execFile("docker", ["cp", `${containerName}:${sandboxPath}`, hostPath], (error) => {
                        if (error) {
                            reject(new Error(`docker cp (out) failed: ${error.message}`));
                        }
                        else {
                            resolve();
                        }
                    });
                }),
                close: async () => {
                    process.removeListener("exit", onExit);
                    process.removeListener("SIGINT", onSignal);
                    process.removeListener("SIGTERM", onSignal);
                    await Effect.runPromise(removeContainer(containerName));
                },
            };
            return handle;
        },
    });
};
// Re-export for backwards compatibility
export { defaultImageName };
const checkImageUid = (imageName, expectedUid) => new Promise((resolve, reject) => {
    execFile("docker", ["image", "inspect", imageName, "--format", "{{.Config.User}}"], (error, stdout) => {
        if (error) {
            reject(new Error(`Image '${imageName}' not found locally. Build it first with 'sandcastle docker build-image'.`));
            return;
        }
        const imageUser = (stdout ?? "").toString().trim();
        if (!imageUser) {
            // No USER directive in image — skip check
            resolve();
            return;
        }
        const uidPart = imageUser.split(":")[0];
        const imageUid = parseInt(uidPart, 10);
        if (isNaN(imageUid)) {
            // Non-numeric user (e.g. "agent") — can't compare, skip check
            resolve();
            return;
        }
        if (imageUid !== expectedUid) {
            reject(new Error(`UID mismatch: image '${imageName}' was built with UID ${imageUid}, ` +
                `but the expected UID is ${expectedUid}. ` +
                `Rebuild the image with 'sandcastle docker build-image', ` +
                `or pass containerUid: ${imageUid} to docker() to match the image.`));
        }
        else {
            resolve();
        }
    });
});
