// @ai-hero/sandcastle/dist/sandboxes/no-sandbox.js
/**
 * No-sandbox provider — runs the agent directly on the host with no container isolation.
 *
 * Usage:
 *   import { noSandbox } from "sandcastle/sandboxes/no-sandbox";
 *   await interactive({ agent: claudeCode("claude-opus-4-6"), sandbox: noSandbox() });
 *
 * Only valid for `interactive()` — not accepted by `run()` or `createSandbox()`.
 * Does not pass `--dangerously-skip-permissions` to the agent — the user manages
 * permissions themselves.
 */
import { spawn } from "node:child_process";
import { createInterface } from "node:readline";
/**
 * Create a no-sandbox provider.
 *
 * The returned provider runs the agent directly on the host. All three
 * branch strategies are supported (head, merge-to-head, branch),
 * defaulting to head.
 */
export const noSandbox = (options) => ({
    tag: "none",
    name: "no-sandbox",
    env: options?.env ?? {},
    create: async (createOptions) => {
        const worktreePath = createOptions.worktreePath;
        const processEnv = { ...process.env, ...createOptions.env };
        const handle = {
            worktreePath,
            exec: (command, opts) => {
                // sudo is a no-op for no-sandbox — the user is already on the host
                const cwd = opts?.cwd ?? worktreePath;
                return new Promise((resolve, reject) => {
                    const proc = spawn("sh", ["-c", command], {
                        cwd,
                        env: processEnv,
                        stdio: [
                            opts?.stdin !== undefined ? "pipe" : "ignore",
                            "pipe",
                            "pipe",
                        ],
                    });
                    if (opts?.stdin !== undefined) {
                        proc.stdin.write(opts.stdin);
                        proc.stdin.end();
                    }
                    const stdoutChunks = [];
                    const stderrChunks = [];
                    if (opts?.onLine) {
                        const rl = createInterface({ input: proc.stdout });
                        rl.on("line", (line) => {
                            stdoutChunks.push(line);
                            opts.onLine(line);
                        });
                    }
                    else {
                        proc.stdout.on("data", (chunk) => {
                            stdoutChunks.push(chunk.toString());
                        });
                    }
                    proc.stderr.on("data", (chunk) => {
                        stderrChunks.push(chunk.toString());
                    });
                    proc.on("error", (error) => {
                        reject(new Error(`exec failed: ${error.message}`));
                    });
                    proc.on("close", (code) => {
                        resolve({
                            stdout: stdoutChunks.join(opts?.onLine ? "\n" : ""),
                            stderr: stderrChunks.join(""),
                            exitCode: code ?? 0,
                        });
                    });
                });
            },
            interactiveExec: (args, opts) => {
                return new Promise((resolve, reject) => {
                    const [cmd, ...rest] = args;
                    const proc = spawn(cmd, rest, {
                        cwd: opts.cwd ?? worktreePath,
                        env: processEnv,
                        stdio: [opts.stdin, opts.stdout, opts.stderr],
                    });
                    proc.on("error", (error) => {
                        reject(new Error(`exec failed: ${error.message}`));
                    });
                    proc.on("close", (code) => {
                        resolve({ exitCode: code ?? 0 });
                    });
                });
            },
            close: async () => {
                // No-op — no container to tear down
            },
        };
        return handle;
    },
});
